site/src/contexts/ProxyContext.tsx

```diff
@@ -0,0 +1,211 @@
+    // and other tabs.
+    savePreferredProxy(preferred)
+    // Set the state for the current context.
+    setProxy(preferred)
+  }
+
+  const queryKey = ["get-regions"]
+  const { error: regionsError, isLoading: regionsLoading } = useQuery({
+    queryKey,
+    queryFn: getWorkspaceProxies,
+    // This onSuccess ensures the local storage is synchronized with the
+    // regions returned by coderd. If the selected region is not in the list,
+    // then the user selection is removed.
+    onSuccess: (data) => {
+      setAndSaveProxy(data.regions, proxy.selectedRegion)
+    },
+  })
+
+  // ******************************* //
+  // ** This code can be removed  **
+  // ** when the experimental is  **
+  // **       dropped             ** //
+  const dashboard = useDashboard()
+  const appHostQueryKey = ["get-application-host"]
+  const {
+    data: applicationHostResult,
+    error: appHostError,
+    isLoading: appHostLoading,
+  } = useQuery({
+    queryKey: appHostQueryKey,
+    queryFn: getApplicationsHost,
+  })
+
+  // If the experiment is disabled, then make the setState do a noop.
+  // This preserves an empty state, which is the default behavior.
+  if (!dashboard?.experiments.includes("moons")) {
+    const value = getURLs([])
+
+    return (
+      <ProxyContext.Provider
+        value={{
+          proxy: {
+            ...value,
+            preferredWildcardHostname:
+              applicationHostResult?.host || value.preferredWildcardHostname,
+          },
+          isLoading: appHostLoading,
+          error: appHostError,
+          setProxy: () => {
+            // Does a noop
+          },
+        }}
+      >
+        {children}
+      </ProxyContext.Provider>
+    )
+  }
+  // ******************************* //
+
+  // TODO: @emyrk Should make an api call to /regions endpoint to update the
+  // regions list.
+
+  return (
+    <ProxyContext.Provider
+      value={{
+        proxy: proxy,
+        isLoading: regionsLoading,
+        error: regionsError,
+        // A function that takes the new regions and selected region and updates
+        // the state with the appropriate urls.
+        setProxy: setAndSaveProxy,
+      }}
+    >
+      {children}
+    </ProxyContext.Provider>
+  )
+}
+
+export const useProxy = (): ProxyContextValue => {
+  const context = useContext(ProxyContext)
+
+  if (!context) {
+    throw new Error("useProxy should be used inside of <ProxyProvider />")
+  }
+
+  return context
+}
+
+/**
+ * getURLs is a helper function to calculate the urls to use for a given proxy configuration. By default, it is
+ * assumed no proxy is configured and relative paths should be used.
+ * Exported for testing.
+ *
+ * @param regions Is the list of regions returned by coderd. If this is empty, default behavior is used.
+ * @param selectedRegion Is the region the user has selected. If this is undefined, default behavior is used.
+ */
+export const getURLs = (
```

Comment:
This isn't really just getting URLs, is it? It's really setting the entire PreferredProxy object.

Reply:
Updated the name